src/versionize_derive/src/descriptors/union_desc.rs

```diff
@@ -0,0 +1,174 @@
+
+use common::{Descriptor, Exists, FieldType, GenericDescriptor};
+use fields::union_field::*;
+use helpers::{compute_version, generate_deserializer_header};
+use quote::quote;
+
+pub(crate) type UnionDescriptor = GenericDescriptor<UnionField>;
+
+impl Descriptor for UnionDescriptor {
+    fn generate_serializer(&self) -> proc_macro2::TokenStream {
+        let mut versioned_serializers = proc_macro2::TokenStream::new();
+
+        for i in 1..=self.version {
+            let union_serializer = self.generate_union_serializer(i);
+
+            versioned_serializers.extend(quote! {
+                #i => {
+                    #union_serializer
+                }
+            });
+        }
+
+        versioned_serializers
+    }
+
+    fn generate_deserializer(&self) -> proc_macro2::TokenStream {
+        let mut versioned_deserializers = proc_macro2::TokenStream::new();
+
+        for i in 1..=self.version {
+            let union_serializer = self.generate_union_deserializer(i);
+
+            versioned_deserializers.extend(quote! {
+                #i => {
+                    let mut object = Self::default();
+                    #union_serializer;
+                    Ok(object)
+                }
+            });
+        }
+
+        quote! {
+            let version = version_map.get_type_version(app_version, Self::type_id());
+            match version {
+                #versioned_deserializers
+                _ => panic!("Unknown {:?} version {}.", Self::type_id(), version)
+            }
+        }
+    }
+
+    fn version(&self) -> u16 {
+        self.version
+    }
+
+    fn ty(&self) -> String {
+        self.ty.to_string()
+    }
+}
+
+impl UnionDescriptor {
+    pub fn new(input: &syn::DataUnion, ident: syn::Ident) -> Self {
+        let mut descriptor = UnionDescriptor {
+            ty: ident,
+            version: 1, // struct start at version 1.
+            fields: vec![],
+        };
+
+        descriptor.parse_union_fields(&input.fields);
+        descriptor.version = compute_version(&descriptor.fields);
+        descriptor
+    }
+
+    fn parse_union_fields(&mut self, fields: &syn::FieldsNamed) {
+        let pairs = fields.named.pairs();
+        for field in pairs.into_iter() {
+            self.fields.push(UnionField::new(self.version, field));
+        }
+    }
+
+    fn generate_field_finder(
+        &self,
+        sizes: proc_macro2::TokenStream,
+        matcher: proc_macro2::TokenStream,
+    ) -> proc_macro2::TokenStream {
+        quote! {
+            // Create a vector of field sizes.
+            let size_vector = vec![#sizes];
+            let mut max: usize = 0;
+            let mut largest_field_index: usize = 0;
+
+            // Find largest field index.
+            for i in 0..size_vector.len() {
+                if (size_vector[i] > max) {
+                    max = size_vector[i];
+                    largest_field_index = i;
+                }
+            }
```

Comment:
nit: slightly more rustacean, but it brings the caveat of having to `use cmp::Ordering` wherever you `derive(Versionize)`:

```suggestion
           let largest_field_index: usize = size_vector
                .iter()
                .enumerate()
                .max_by(|(_, lsize), (_, rsize)| lsize.partial_cmp(rsize).unwrap_or(Ordering::Equal))
                .map(|(index, _)| index)
                .expect("Cannot find largest union field index");
```

Reply:
Thanks